contrib/ros/rikai/contrib/ros/convert.py

```diff
@@ -0,0 +1,348 @@
+        "uint16": IntegerType,
+        "uint32": LongType,
+        "uint64": LongType,
+        "float32": FloatType,
+        "float64": DoubleType,
+        "string": StringType,
+        "time": TimestampType,
+        "duration": LongType,
+        "byte": ByteType,
+        "byte[]": BinaryType,
+        "uint8[]": BinaryType,
+    }
+
+    SCHEMA_CACHE = {}
+
+    def is_supported(
+        self,
+        message_type: str,
+    ) -> bool:
+        if message_type in self.CONVERT_MAP:
+            return True
+        if message_type == "uint64":
+            raise ValueError("uint64 is overflow in Spark")
+
+        # TODO: use local assignment after python 3.8
+        array_type_and_size = parse_array(message_type)
+        if array_type_and_size:
+            return self.is_supported(array_type_and_size[0])
+
+        return False
+
+    def convert(self, message_type: str, value: genpy.message.Message):
+        if message_type in self.CONVERT_MAP:
+            return self.CONVERT_MAP[message_type]()
+        array_type_and_size = parse_array(message_type)
+        if array_type_and_size and self.is_supported(array_type_and_size[0]):
+            return ArrayType(self.convert(array_type_and_size[0], value))
+        return None
+
+    def get_value(self, value, attr_name):
+        msg_type = value._slot_types[value.__slots__.index(attr_name)]
+        if msg_type in self.SCHEMA_CACHE:
+            return self.SCHEMA_CACHE[msg_type]
+        if self.is_supported(msg_type):
+            return msg_type
+        if Visitor.parse_array(msg_type):
+            return msg_type
+        modules = msg_type.split("/")
+        try:
+            mod = importlib.import_module(".".join(modules[:-1] + ["msg"]))
+            return getattr(mod, modules[-1])
+        except ModuleNotFoundError:
+            logging.error(
+                "Can not find module: %s, %s, %s", value, attr_name, msg_type
+            )
+            raise
+
+    def build_record(self, fields):
+        return StructType([StructField(k, v) for k, v in fields.items()])
+
+
+class Visitor:
+    """Visitor walks through a ROS Message and converts it to
+    another type system (i.e., Spark or Numpy).
+    """
+
+    def __init__(self, converter: Converter) -> None:
+        self.converter = converter
+
+    @staticmethod
+    def message_fields(
+        message: genpy.message.Message,
+    ):
+        return zip(message.__slots__, message._slot_types)  #
+
+    @staticmethod
+    def parse_array(message_type: str) -> Optional[Tuple[str, Optional[int]]]:
+        """Parse array defination.
+
+        Parameters
+        ----------
+        message_type : str
+            ROS Message type, in str, i.e., `std_msgs/ByteMultiArray`.
+
+        Return
+        ------
+        A tuple of `[element type, array_length]` if it is an fixed-size array.
+        A tuple of `[element type, None]` if it is variable-length array.
+        Return None if it is not an array.
```

Comment:
nit: it is better to return a tuple of two values for any type? For non-array, just return None, None. So that when calling this func you can always do

```
type, size = parse_array(val)

If type:
   ...
```

Reply:
Good point. will do. thanks!